### DISCUSSION TITLE
Adding Brands to products

### DISCUSSION THREAD

**Discussion by jacobkerber** [medusajs/medusa#6011] [Q&A] [answered]
Is there any ability to add brands to a product? For example, I can get the same SKU from multiple brands.

**Comment by olivermrbl** [comment, marked as answer]
We don't support brands on products natively. Would [product tags](https://docs.medusajs.com/api/admin#product-tags_getproducttags) suffice for your use case? 

Alternatively, you could [create a separate data model](https://docs.medusajs.com/development/entities/create), ProductBrand, that holds references to products. That would give you a more flexible solution in case you need to expand on the concept of a product brand in the future.

**Reply by jacobkerber** [reply]
@olivermrbl Thank you for the info. I don't think product tags would work as I think I would also like to have brand management in the admin panel. I could start the adoption of Medusa using tags to assign brands.

In response to the data model approach, I think this is the best option. I am relatively new to node, haven't touched Postgresql before, and haven't look at JS in a long time. I will start to look at the documentation to get his together. I think the hardest part for mewl be integrating it into the admin panel.

**Reply by olivermrbl** [reply]
Let me know if I can help.

We have a strong toolbox for building custom functionality, including data models and admin panel extensions. I will attach a couple of resources (in addition to the link for data models above) that might be relevant to your exploration:
- [Creating services](https://docs.medusajs.com/development/services/create-service). This is typically where the business logic lives.
- [Creating API routes/endpoints](https://docs.medusajs.com/development/api-routes/create). Exposing the functionality of your services through a Rest API.
- [Admin UI routes](https://docs.medusajs.com/admin/routes). Blank canvas to create extensions from.
- [Admin UI Widgets](https://docs.medusajs.com/admin/widgets). Custom components on existing pages in the admin panel.

**Reply by jacobkerber** [reply]
@olivermrbl What entity should I extend in my custom data model? Is there a list of extensible options? Once I have my model, how can I add a field in the admin UI for either the variant or product to add the Brand name into?

**Reply by imash96** [reply]
To add brand field in Product you can do the following things:

first extend model of core entity Product
.\src\models\product.ts

```
import { Entity, Column } from "typeorm"
import { Product as MedusaProduct } from "@medusajs/medusa"

@Entity()
export class Product extends MedusaProduct {
    @Column({ type: "varchar", length: 100, nullable: true })
    customattribute: string | null

    @Column({ type: "varchar", length: 50, nullable: true })
    brand: string | null

    @Column({ type: "varchar", length: 10, nullable: true })
    gender: string | null

    @Column({ type: "varchar", length: 50, nullable: true })
    care: string | null

    @Column({ type: "varchar", length: 50, nullable: true })
    style: string | null
}
```
Then create migration file as below
./src/migrations/1710754068334-productCustomization.ts

```
import { MigrationInterface, QueryRunner } from "typeorm";

export class ProductCustomization1710754068334 implements MigrationInterface {
    name = 'ProductCustomization1710754068334'

    public async up(queryRunner: QueryRunner): Promise<void> {
        await queryRunner.query(`ALTER TABLE "product" ADD "customattribute" character varying(100)`);
        await queryRunner.query(`ALTER TABLE "product" ADD "brand" character varying(50)`);
        await queryRunner.query(`ALTER TABLE "product" ADD "gender" character varying(10)`);
        await queryRunner.query(`ALTER TABLE "product" ADD "care" character varying(50)`);
        await queryRunner.query(`ALTER TABLE "product" ADD "style" character varying(50)`);
    }
    public async down(queryRunner: QueryRunner): Promise<void> {
        await queryRunner.query(`ALTER TABLE "product" DROP COLUMN "style"`);
        await queryRunner.query(`ALTER TABLE "product" DROP COLUMN "care"`);
        await queryRunner.query(`ALTER TABLE "product" DROP COLUMN "gender"`);
        await queryRunner.query(`ALTER TABLE "product" DROP COLUMN "brand"`);
        await queryRunner.query(`ALTER TABLE "product" DROP COLUMN "customattribute"`);
    }
}
```
Now add loader file to include these fileds in response json
./src/loaders/extend-product-fields.ts
```
export default async function () {
    const storeProductImports = (await import(
        "@medusajs/medusa/dist/api/routes/store/products/index"
    )) as any
    const adminProductImports = (await import(
        '@medusajs/medusa/dist/api/routes/admin/products/index'
    )) as any;
    storeProductImports.defaultStoreProductsFields = [
        ...storeProductImports.defaultStoreProductsFields,
        "customattribute",
        "brand",
        "gender",
        "care",
        "style",
    ]
    // admin
    adminProductImports.defaultAdminProductsFields = [
        ...adminProductImports.defaultAdminProductsFields,
        "customattribute",
        "brand",
        "gender",
        "care",
        "style",
    ]
}
```
These are the basic steps to extend the core entity in backend. Now what you can do is.
1) Extend the validator which validate the post request to create and update the product.
[https://docs.medusajs.com/development/api-routes/extend-validator](url)

2) You can create the ui in admin dashboard to include this fileds.
[https://docs.medusajs.com/admin/widgets](url)
[https://docs.medusajs.com/admin/routes](url)

If you care facing issue in the above two step what you can do is create a custom route and service. In service you can write code to update the product with brand details then pass the brand details in post request in browser console.
[https://docs.medusajs.com/development/api-routes/create](url)
[https://docs.medusajs.com/development/services/create-service](url)

**Comment by darrylmorley** [comment]
@olivermrbl Please don't take this the wrong way but I'm really interested why Products don't already have a brand? It just feels like a really strange omission to me. Are most users using the platform to sell items from one brand only? All products have a brand, often it can be an important consideration for buyers when making a purchase. It's just weird 🤣. Medusa aren't the only ones to omit this, it seems to be more common in newer frameworks.

**Reply by thubamamba** [reply]
This is literally my frustration right now.. I am down for adopting Medusa for a project but the hoops to just add Brand functionality? That is tedious for a project that's at this stage. 

> @olivermrbl Please don't take this the wrong way but I'm really interested why Products don't already have a brand? It just feels like a really strange omission to me. Are most users using the platform to sell items from one brand only? All products have a brand, often it can be an important consideration for buyers when making a purchase. It's just weird 🤣. Medusa aren't the only ones to omit this, it seems to be more common in newer frameworks.
